### PR TITLE
feat: support user-defined formats

### LINKED ISSUE #1: extend support to non-json files
Hi, we are using **meta-updater** for some json files, and this is really powerful and convenient tool, however we are limited to only files with `.json` extension.

We would like to also process the following files in meta-updater:
- `.npmrc`, `.gitignore`, `.browserlistrc` (text content)
- `somefile.cjs`, `prettier.config.js` (js content)
- `.gitlab-ci.yaml` (yaml content)

### JavaScript

And we will definitely know the format for each file, so the simplest change would be to replace the following:
https://github.com/pnpm/meta-updater/blob/4dde93ee877a9210e05ec60684a0f9aac09aa763/src/index.ts#L56-L57

with:
```ts
if (!p.endsWith('.json')) continue
const obj = p.endsWith('.json') ? await readJsonFile(fp) : await readFile(fp, 'utf8')
```
and add `import { readFile } from 'fs/promises'` to the top of file

### Type Checking 
Also `update` parameter can be strongly typed depending if file has or hasn't `.json` extension: [ts playground](https://www.typescriptlang.org/play?#code/PTAEEsFsAcHsCcAuBnUA6DBYAUIgntAKagAK8sAVoQMaICyAhgHbgBmhyioAvKAN4BfANw4QoHBNwFiAYXiEGiQgFVoAE0WEAYgFcm1ADwAVAHw9QAClgAjCgC5QRgDSg14eA87xwTAOYvIZjYORAcyShp6IPZOAEoeMyNQAB9SckhwZEJjFNAmHQAbArNU-KLJfCJQVQ0lAFF4cnhzPhxQUEIADyJaQjUHAHlbSNyvH19csoK20AZaHQYCweHaUcRvP0nC6ex26EUAC0918ZwBSS64JFnkPH1QVj1acFgmUCJ4VgRIGs1kHK6SiYalQACUaAg1AYxn4XHoANZMWAAdyYJhMFhmyIQ8OQ+2ohAAIu5jht-DMdOpNA4+KAANoAaQgb3hhDwsFYjgAug4mYDCMDUAADAAkfBhvgEaAoyFeQtAAH5QHIFEpfkpdPoDDYqLQzA4VZp1dontCTn4zAInDNYNAUAqaaAlJwHaBrLBYAUFG9zthYmF0plslNcsaGk0zK1dqAxBg0DN5IgdPA3iGGKhmHgRNhfTgPl94D8qc7MdGAOQyOwAHXCupQVegTGgkDL1ujUfa7TL+PhDF8hGlsqYZYctQYCTy2xcYgOhHkrkU4-28AYkEISmazDUoETybeADdFjpiAx55ViNRZ9RWWoZl20E3IPBqCObndqE7AROlJ1ENOwLO85ju8p6ruuc6zMCO7rnuoCHgUx6zGe0igJeNA3jMvqxEAA)

### PR DESCRIPTION
**Pull request by ibezkrovnyi**
resolves #1

- usage examples are in 'src/examples' directory
- format is associated with extension. File extension can be set explicitly with square brackets or if there are no square brackets - extension is taken from file
  - `package.json` - detected '.json' extension 
  - `.eslintrc [.json]` - explicitly set '.json' extension
- supports 3 options formats:

### simplified
```js
{
    ".eslintrc [.json]": (content, options) => ...
}
```
### explicit with 'files' property
```js
{
  "files": {
    "package.json": (content, { dir, manifest, file, resolvedPath, _writeProjectManifest }) => ...
  }
}
```
### with user-defined formats
```js
{
  "files": {
    ".somefile [.csv]": () => ...
  },
  "formats": {
    ".csv": createFormat({ ... })
  }
}
```

### CODE REVIEW

File: src/updater/formatPlugin.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+  /** Called only if check for equality is required (`actual != null` & `expected != null`) */
+  equal(expected: Content, actual: Content, options: FormatPluginFnOptions): PromiseOrValue<boolean>
+
+  /** Called only if write is required (`--test` isn't specified, `expected != null` and `expected` is not equal to `actual`) */
+  write(expected: Content, options: FormatPluginFnOptions): PromiseOrValue<void>
+}
+
+export interface FormatPluginFnOptions {
+  // checkOnly: boolean
```
Comment (ibezkrovnyi) [inline src/updater/formatPlugin.ts:30]: bcs now options are passed as object, we can pass some boolean variable here e.g. `checkOnly` or `shouldUpdate` that reflects `opts?.test`

File: tsconfig.json [hunk 2]
```diff
@@ -9,10 +9,14 @@
...
     "strictNullChecks": true,
+    "lib": ["ES2020"],
+    "types": ["node", "jest"],
     "target": "es2019",
     "outDir": "lib",
-    "module": "ES2020",
-    "moduleResolution": "node"
+    "module": "es2020",
+    "moduleResolution": "node",
+    "skipLibCheck": true,
```
Comment (ibezkrovnyi) [inline tsconfig.json:18]: issues with some `@yarnpkg` type

File: .prettierrc.json [hunk 3]
```diff
@@ -0,0 +1,7 @@
+{
+  "semi": false,
+  "singleQuote": true,
+  "useTabs": false,
+  "tabWidth": 2,
+  "printWidth": 120
```
Comment (ibezkrovnyi) [inline .prettierrc.json:6]: you may want to remove this file, but then prettier in vscode/intellij will format with defaults

File: README.md [hunk 4]
```diff
@@ -37,6 +40,58 @@ export default (workspaceDir) => {
...
+      '.prettierignore [#ignore]': (actual) => actual ?? ['node_modules'],
+    },
+    formats: {
+      '#ignore': ignoreFormat,
+    },
+  })
+}
+```
+
+See more examples at [src/examples/](src/example/)
```
Comment (ibezkrovnyi) [inline README.md:93]:
```suggestion
See more examples at [src/examples/](src/examples/)
```

File: README.md [hunk 5]
```diff
@@ -37,6 +40,58 @@ export default (workspaceDir) => {
 }
 ```
 
+### Example 2
+
+The following `.meta-updater/main.mjs` defines format `#ignore` and updaters for `.gitignore` and `.eslintrc` using explicit format specifier.
+
+```js
+import * as fs from 'fs/promises'
+import { createFormat, createUpdateOptions } from '@pnpm/meta-updater'
```
Comment (zkochan) [inline README.md:49]: Why do you need these functions?

You could just export a new field from the update file. Like

```
export const formats = {
   ...
}
```

Comment (ibezkrovnyi) [reply README.md:49]: Both functions are used to constraint types.

## Ways to use meta-updater

#### Unsafe: JavaScript without IDE/without typescript language server

No need to use `createFormat` / `createUpdateOptions` at all. You can see it in action in `__fixtures__/workspace-1/.meta-updater/main.mjs` - it is plain object without `createUpdateOptions`.

#### Unsafe: JavaScript with IDE/with typescript language server

Without manually annotating with JSDocs and without `createFormat` / `createUpdateOptions` all the parameters in file updaters will be inferred as `any`. Nothing stops from using meta-updater like this.

#### Safe: TypeScript or JavaScript with IDE/with typescript language server

Everything should be annotated with types (TS) or JSDocs (JS) or `createFormat` / `createUpdateOptions` should be used.

## Format approaches - Safe only

#### Without `createFormat` in TypeScript

pros:
- no use of createFormat / createUpdateOptions

cons:
- need to annotate everything with types every time for every user (not DRY)
- mental load - need to know where content is of type `string[]` and where it is of type `string[] | null`

```ts
const gitignoreFormat = {
  async read({ resolvedPath }: { dir: string, manifest: import('@pnpm/types'). ProjectManifest, resolvedPath: string }) {
    return (await fs.readFile(resolvedPath, 'utf8')).split('\n')
  },
  update(actual: string[] | null, updater: (content: string[] | null, options: { dir: string, manifest: import('@pnpm/types'). ProjectManifest, resolvedPath: string }) => null | string[] | Promise<null | string[]>, options: { dir: string, manifest: import('@pnpm/types'). ProjectManifest, resolvedPath: string }) {
    return updater(actual, options)
  },
  equal(expected: string[], actual: string[]) {
    return equals(expected, actual)
  },
  async write(expected: string[], { resolvedPath }: { dir: string, manifest: import('@pnpm/types'). ProjectManifest, resolvedPath: string }) {
    const unique = <T extends unknown[]>(array: T) => Array.from(new Set<T[number]>(array)).sort()
    await fs.writeFile(resolvedPath, unique(expected).join('\n'), 'utf8')
  },
}
```

#### Without `createFormat` in JavaScript (same pros and cons as the above)

pros:
- no use of createFormat / createUpdateOptions

cons:
- need to annotate everything with types every time for every user (not DRY)
- mental load - need to know where content is of type `string[]` and where it is of type `string[] | null`

```js
const gitignoreFormat = {
  /**
   * @param {{dir: string, manifest: import('@pnpm/types').ProjectManifest}} param0 
   */
  async read({ resolvedPath }) {
    return (await fs.readFile(resolvedPath, 'utf8')).split('\n')
  },
  /**
   * @param {string[] | null} actual
   * @param {{(content: string[] | null, options: {dir: string, manifest: import('@pnpm/types').ProjectManifest}): null | string[] | Promise<null | string[]>}} updater
   * @param {{dir: string, manifest: import('@pnpm/types').ProjectManifest}} param2 
   */
   update(actual, updater, options) {
    return updater(actual, options)
  },
  /**
   * 
   * @param {string[]} expected 
   * @param {string[]} actual 
   * @returns 
   */
  equal(expected, actual) {
    return R.equals(expected, actual)
  },
  /**
   * @param {string[]} expected
   * @param {{dir: string, manifest: import('@pnpm/types').ProjectManifest}} param1 
   */
   async write(expected, { resolvedPath }) {
    const unique = (array) => Array.from(new Set() < T[number] > array).sort()
    await fs.writeFile(resolvedPath, unique(expected).join('\n'), 'utf8')
  },
}
```

#### Format definition with `createFormat`:

pros:
- works in both TypeScript and JavaScript with zero type annotations
- type of format content is inferred from return type of the `read` method and contextually typed in all other locations:
- types of parameters are contextually typed

cons:
- runtime overhead for basically identity function (createFormat compiled to JS looks like: `const createFormat = a => a`)

```js
const gitignoreFormat = createFormat({
  async read({ resolvedPath }) {
    return (await fs.readFile(resolvedPath, 'utf8')).split('\n')
  },
  update(actual, updater, options) {
    return updater(actual, options)
  },
  equal(expected, actual) {
    return equals(expected, actual)
  },
  async write(expected, { resolvedPath }) {
    const unique = (array) => Array.from(new Set(array)).sort()
    await fs.writeFile(resolvedPath, unique(expected).join('\n'), 'utf8')
  },
})
```

## How type safety looks (screenshots are from JavaScript main.mjs):

#### files
image

image

image

image

image

#### format

image

image

**Review by zkochan — APPROVED** [review]

**Comment by zkochan** [comment]
I am not sure why you have bumped the version in the PR. Now I released 0.2.0

**Comment by ibezkrovnyi** [comment]
> I am not sure why you have bumped the version in the PR. Now I released 0.2.0

Apologize for the accident, I assume if there is no configured `changesets` in the package.json#scripts, version bump has to be done manually.

**Pull request merged**